### PR TITLE
Add ChannelKeys to ChannelMonitor

### PR DESCRIPTION
**Pull request by devrandom**
This is the first step to removing secret key handling from `ChannelMonitor`.

Further steps will include:

- moving left over signing operations into `ChannelKeys`
- eliminating the secret keys

I broke this into steps in order to reduce PR size, even though it does introduce some temporary redundancy.

BTW, I find the generic on `ChanSigner` to be somewhat painful.  It seems that the main reason for this is to have fast inline serialization.  Maybe there's a way to do that while keeping the `ChanSigner` dynamic.

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -372,8 +372,7 @@ pub(super) enum ChannelError {
 	Ignore(&'static str),
 	Close(&'static str),
 	CloseDelayBroadcast {
-		msg: &'static str,
-		update: Option<ChannelMonitor>
```
Comment (devrandom) [inline lightning/src/ln/channel.rs]: Seems that this field wasn't needed, because the object is available at the point where this error is handled.

**Comment by ariard** [comment]
@devrandom maybe have a high-level look on #462, it's a refactor to make tx generation in one place and split it from txn parsing. 

What would you expect as an nterface for ChanSigner concerning non-BOLT3-specified transactions like justice tx on revoked commitment or spending remote HTLC outputs ? 

```
fn sign_justice_tx(Vec<RevokedOutpointDescriptor) {}

RevokedHTLCOutpointDescriptor {
        commitment_number: u32,
        hash: PaymentHash
        lock_time: u32
}

RevokedOutpointDescriptor {
        commitment_number: u32,
        csv_delay: u32,
```

Assuming Signer have knowledge of redeemScript, preimages, revocation secrets. So roughly it would a SpendableOuputDescripor extension but would be better to use some well-defined data structures for inter-operability like PSBT (or pass some output script descriptor and let the signer implementor generate the PSBT if there is multiple signers ?)

File: lightning/src/ln/channelmonitor.rs [hunk 2]
```diff
@@ -726,12 +727,255 @@ impl PartialEq for ChannelMonitor {
 	}
 }
 
-impl ChannelMonitor {
-	pub(super) fn new(funding_key: &SecretKey, revocation_base_key: &SecretKey, delayed_payment_base_key: &SecretKey, htlc_base_key: &SecretKey, payment_base_key: &SecretKey, shutdown_pubkey: &PublicKey, our_to_self_delay: u16, destination_script: Script, logger: Arc<Logger>) -> ChannelMonitor {
+impl<ChanSigner: ChannelKeys + Writeable> ChannelMonitor<ChanSigner> {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Hmm, *moving* code hunks in the same commit where they are changed makes it super hard to review. If you want to move a code hunk, best to do it in its own commit clearly tagged move-only so that its easy to verify.

File: lightning/src/ln/channel.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs]: Hmm, fine for now, I suppose, eventually it should be an opaque update object, but tht'll be after #387.

File: lightning/src/chain/keysinterface.rs [hunk 3]
```diff
@@ -193,19 +195,56 @@ pub struct InMemoryChannelKeys {
...
+	pub local_channel_pubkeys: ChannelPublicKeys,
 	/// Remote public keys and base points
 	pub remote_channel_pubkeys: Option<ChannelPublicKeys>,
 	/// The total value of this channel
 	pub channel_value_satoshis: u64,
 }
 
+impl InMemoryChannelKeys {
+	/// Create a new InMemoryChannelKeys
+	pub fn new<C: Signing>(
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs]: If you're gonna have a new, probably best to hide the internal fields, no?

File: lightning/src/chain/keysinterface.rs [hunk 4]
```diff
@@ -133,7 +133,7 @@ pub trait KeysInterface: Send + Sync {
 /// (TODO: We shouldn't require that, and should have an API to get them at deser time, due mostly
 /// to the possibility of reentrancy issues by calling the user's code during our deserialization
 /// routine).
-pub trait ChannelKeys : Send {
+pub trait ChannelKeys : Send+Clone+PartialEq {
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs]: Oof. Clone sucks, especially since the enforcement then cant be in our own ChannelKeys but has to be some global hashmap tracking. I suppose its largely something that can only be resolved with #387, so if you just add a comment noting that we need to remove this ASAP, maybe that's OK. Also, can we drop the PartialEq and have the ChannelMonitor PartialEq compare pubkeys()?

File: lightning/src/ln/channelmonitor.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: I had no choice, because of the need for `+ Writeable`.  But you are right, I should have split that to its own commit.  I'll look into doing that.

**Comment by devrandom** [comment]
@ariard @TheBlueMatt , thank you for bringing #461 / #462 to my attention, I should have looked before getting started.  I am totally fine with rebasing this as needed.

File: lightning/src/ln/channel.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/ln/channel.rs]: OK

File: lightning/src/chain/keysinterface.rs (hunk 3 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Done, but a couple of `pub(crate)` fields due to reference from the enforcing impl

File: lightning/src/chain/keysinterface.rs (hunk 4 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Done, on the Storage enum

**Comment by devrandom** [comment]
@ariard:

> Assuming Signer have knowledge of redeemScript, preimages, revocation secrets. So roughly it would a SpendableOuputDescripor extension but would be better to use some well-defined data structures for inter-operability like PSBT (or pass some output script descriptor and let the signer implementor generate the PSBT if there is multiple signers ?)

For closing/justice inputs, each of them is determined by protocol, and the information needed is as you have in your comment AFAICT.  I think PSBT is not very suitable for describing the inputs, because much of the information is Lightning specific.

For the outputs, they can be anything, and they could in theory be described by PSBT or some other mechanism.  I think that's out of scope for lightning-rust, at least for now.  The outputs will of course be examined by the signer to ensure they follow policy (e.g. go to a whitelisted destination).

Also, for rust-lightning, since it's a development kit, it should not concern itself with how the outputs are constructed, but just give the developer freedom to choose.

Did I answer your question?

I'll look at #462 in more detail.

File: lightning/src/ln/channelmonitor.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: Also related is my comment above that ChanSigner would benefit from being dynamic.

**Comment by devrandom** [comment]
@ariard #462 looks like a much needed refactoring to reduce the complexity of ChannelMonitor.

Please let me know if you want me to comment on anything specific.

**Comment by ariard** [comment]
@devrandom you may use PSBT even for LN, by passing per_commitment in BIP_IN_BIP32_DERIVATION and witnessScript in PSBT_IN_WITNESS_SCRIPT, assuming you also send revocation_secret/preimage to signer. 

> Also, for rust-lightning, since it's a development kit, it should not concern itself with how the outputs are constructed, but just give the developer freedom to choose.

I agree, we shouldn't enforce any binary format on the external signer implementor while at the same time avoid a messy custom-data structure to describe outputs.

> For the outputs, they can be anything, and they could in theory be described by PSBT or some other mechanism.

I was thinking about the _in-protocol_ outputs, the ones you _must_ take actions on before a given time, i.e revoked outputs or commitment HTLC outputs. Not outputs belonging to justice/sweeping txn, this ones are completely up to the signer.

Yes thanks for your answer, it will be easier to see if everything suit well when we remove private keys from ChannelMonitor.
 
I'm waiting Matt  high-level feedback on #462 before to build on top and yes remove private keys completely. It may drop some parts of your PR, but that would be for the best I guess ?

**Comment by devrandom** [comment]
> @devrandom you may use PSBT even for LN, by passing per_commitment in BIP_IN_BIP32_DERIVATION and witnessScript in PSBT_IN_WITNESS_SCRIPT, assuming you also send revocation_secret/preimage to signer.

A signer that understands PSBT but doesn't understand Lightning would not be able to sign, right?  So although it's a use of PSBT, it's not standard, and PSBT doesn't solve the entire transmission requirements.

> > For the outputs, they can be anything, and they could in theory be described by PSBT or some other mechanism.
> 
> I was thinking about the _in-protocol_ outputs, the ones you _must_ take actions on before a given time, i.e revoked outputs or commitment HTLC outputs. Not outputs belonging to justice/sweeping txn, this ones are completely up to the signer.

Yes, agreed, the protocol outputs handling has to follow a specific pattern.

> Yes thanks for your answer, it will be easier to see if everything suit well when we remove private keys from ChannelMonitor.
> 
> I'm waiting Matt high-level feedback on #462 before to build on top and yes remove private keys completely. It may drop some parts of your PR, but that would be for the best I guess ?

Yes, of course.

File: lightning/src/ln/channelmonitor.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmonitor.rs]: Hmm? I've never heard of Rust caring about ordering of loose hunks in a file, even if they rely on trait impls defined later in that file?

**Comment by TheBlueMatt** [comment]
@ariard given #462 may take a bit, does it make sense for this to go first?

File: lightning/src/ln/channelmonitor.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmonitor.rs]: Oh, oops, I missed what you did there. So I'm confused, why does ChannelMonitor need to be Writeable now?

Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: This section is now under `impl<ChanSigner: ChannelKeys + Writeable>`, whereas the previous section is now under `impl<ChanSigner: ChannelKeys>`.  I tried putting everything under `+ Writeable`, but the compiler gets unhappy.

Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: Oh, our comments crossed.  Only the `write*` methods need the `+ Writeable` on `ChanSigner` - so that they can serialize the `keys` field.

Comment (TheBlueMatt) [reply lightning/src/ln/channelmonitor.rs]: Ahhh, ok, right, now things make more sense. In any case, can you split the commit (you probably want git reset HEAD^1 && git commit -p or so) so that you first just have two impl ChannelMonitor blocks with the write above (or maybe below where the read is is better?) and then have the second commit that makes changes.

Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: Will do.

Comment (devrandom) [reply lightning/src/ln/channelmonitor.rs]: The split is done.  `git show --color-moved HEAD^`

File: lightning/src/ln/channel.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs]: Oops, right, so I missed that this is largely as an indication as to whether or not we need to do an update-add, less to provide it. Unless there's a reason you need to drop it for this commit, we should back this change out.

Comment (devrandom) [reply lightning/src/ln/channel.rs]: Would it be OK to have it as a bool then?  otherwise, `ChannelError` gets a generic type param and that's a bunch of clutter in the rest of the file.

Comment (devrandom) [reply lightning/src/ln/channel.rs]: Also noticing that it's only constructed in one place.

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs]: Hmm, they're just internal types, having to template them in the return types of many Channel functions doesn't seem to bad. At a high level its pretty awkward to access anything in Channel after it returns a ChannelError::Close, since at that point, it is assumed that the Channel is in an invalid state. Its *probably* fine in nearly every case, but we should avoid it where possible.

File: lightning/src/chain/keysinterface.rs [hunk 5]
```diff
@@ -268,6 +308,7 @@ impl_writeable!(InMemoryChannelKeys, 0, {
 	delayed_payment_base_key,
 	htlc_base_key,
 	commitment_seed,
+	local_channel_pubkeys,
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Done

File: lightning/src/ln/channel.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/ln/channel.rs]: OK, the revert is done.  Let me know once the new commits are reviewed and I'll squash.

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -2196,7 +2196,7 @@ impl<ChanSigner: ChannelKeys, M: Deref> ChannelManager<ChanSigner, M> where M::T
 					return Err(MsgHandleErrInternal::send_err_msg_no_close("Got a message for a channel from the wrong node!", msg.channel_id));
 				}
 				if (msg.failure_code & 0x8000) == 0 {
-					try_chan_entry!(self, Err(ChannelError::Close("Got update_fail_malformed_htlc with BADONION not set")), channel_state, chan);
+					try_chan_entry!(self, Err(ChannelError::<ChanSigner>::Close("Got update_fail_malformed_htlc with BADONION not set")), channel_state, chan);
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs]: 1.22 is sad here (and the next hunk). I think you may have to just do eg
```
					let channel_error: ChannelError<ChanSigner> = ChannelError::Close("Got update_fail_malformed_htlc with BADONION not set");
					try_chan_entry!(self, Err(channel_error), channel_state, chan);
```

**Comment by TheBlueMatt** [comment]
Other than the 1.22 error, looks good!

File: lightning/src/ln/channelmanager.rs (hunk 6 above)
Comment (devrandom) [reply lightning/src/ln/channelmanager.rs]: Fixed.

**Comment by devrandom** [comment]
Squashed to two commits, let me know if you want just one.

File: lightning/src/chain/keysinterface.rs [hunk 7]
```diff
@@ -182,21 +187,58 @@ pub trait ChannelKeys : Send {
...
+	pub fn new<C: Signing>(
+		secp_ctx: &Secp256k1<C>,
+		funding_key: SecretKey,
+		revocation_base_key: SecretKey,
+		payment_base_key: SecretKey,
+		delayed_payment_base_key: SecretKey,
+		htlc_base_key: SecretKey,
+		commitment_seed: [u8; 32],
+		channel_value_satoshis: u64,
+		remote_channel_pubkeys: Option<ChannelPublicKeys>) -> InMemoryChannelKeys {
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs]: Hmm, this is awkward - its pub and if a user ever sets it to Some() for any reason we'll almost certainly panic() in set_remote_channel_pubkeys pretty quick. You should be able to just bypass the new() call in the deserialize and construct manually.

Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Ah, you are right.  Fixed.

File: lightning/src/ln/channelmanager.rs [hunk 8]
```diff
@@ -516,11 +516,10 @@ macro_rules! try_chan_entry {
 					$channel_state.short_to_id.remove(&short_id);
 				}
 				if let Some(update) = update {
-					if let Err(e) = $self.monitor.add_update_monitor(update.get_funding_txo().unwrap(), update) {
+					if let Err(e) = $self.monitor.add_update_monitor(update.get_funding_txo().unwrap(), update.clone()) {
 						match e {
 							// Upstream channel is dead, but we want at least to fail backward HTLCs to save
-							// downstream channels. In case of PermanentFailure, we are not going to be able
-							// to claim back to_remote output on remote commitment transaction. Doesn't
+							// downstream channels. In case of PermanentFailure, we are not going to be able hello this is Miron Cupermanipoipwieeeieie 							// to claim back to_remote output on remote commitment transaction. Doesn't
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs]: Hello Miron Cuperman, this is Matt Corallo?

Comment (devrandom) [reply lightning/src/ln/channelmanager.rs]: gah, fixed
